text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,626 @@
+edge locations close to any customer, and binding together the static assets
+that compose the application and the data objects maintained by the back-end
+pipeline.
+
+1. The landing page asynchronously loads the `latest.json` object to render the
+   "latest" tiles, containing the short-list of packages that changed most
+   recently.
+
+1. The search feature asynchronously loads the `catalog.json` object, and
+   performs client-side filtering to prepare search results.
+
+1. Package detail pages fetch the relevant `assembly.json` object from the
+   assembly store, and renders the detail page using information contained
+   therein.
+
+   - If a different language (than JavaScript) is selected, the
+     `assembly-${lang}.json` object is fetched instead, providing the relevant
+     documentation elements. In the event that object is not available yet, a
+     message will be displayed instead, instructing the user to try again later.
+
+   URL                                                       | Description
+   ----------------------------------------------------------|----------------------------------------------------------
+   `<root>/packages[/@<scope>]/<name>`                       | Redirects to the latest known version of `[@scope/]name`
+   `<root>/packages[/@<scope>]/<name>/v/<version>`           | The overview of `[@scope/]name` at `version`
+   `<root>/packages[/@<scope>]/<name>/v/<version>/docs`      | The documentation (`README.md`) of `[@scope/]name` at `version`
+   `<root>/packages[/@<scope>]/<name>/v/<version>/docs/<FQN>`| The documentation for type `FQN` in `[@scope/]name` at `version`
+
+   - When a `<version>` parameter is only a prefix of a SemVer version number,
+     the URL redirects to the latest known version in that line (e.g: `/v/1`
+     would redirect to `/v/1.2.3`, and `/v/1.1` would redirect to `/v/1.1.9`)
+   - A query parameter (e.g: `?lang=go`) may be added to anchor a documentation
+     page (under the `/docs` scope) on a particular language.
+
+[react]: https://reactjs.org
+
+The details of URLs to be configrued in CloudFront is available in [appendix
+"CloudFront URLs"](#cloudfront-urls).
+
+#### Website Analytics
+
+An analytics solutions will be designed to enable monitoring the success and
+use-cases of the platform. A stream of URLs hit by users will be collected,
+including search terms used by users. That data will be completely anonymous,
+and an opaque identifier will be used to enable approximating "unique visitor"
+counts, etc. No cross-site tracking will be performed.
+
+If IP addresses are fed into the analyitics service, these will not be
+persisted, but may instead be mapped to a coarse-grained location (e.g:
+country).
+
+Legal review will be necessary to determine if the dataset implies compliance
+requirements for GDPR, CCPA, and similar consumer privacy regulations.
+
+#### Construct Packaging
+
+The Construct Hub is to be created as a reusable construct, including all
```

Comment:
## How much does it cost to operate a ConstructHub?

Since we are going to offer ConstructHub as a reusable construct for customers to operate their own instance, we need to have a mechanism to measure the cost of operating a ConstructHub.  Before committing to operate their own version of ConstructHub, customers will want to know it costs as a dependency of different factors such as number of packages ingested, the rate in which packages are ingested and more.

Reply:
Interesting! There are different aspects to cost though (there is infrastructure cost, operational overhead, etc...). It may be difficult to get a sense of these costs before we actually operate our own hub for a short while. I'll nevertheless add a mention to this in the document, because this will eventually be important information!

Reply:
Super important and I think this is an opportunity to innovate and come up with some general model/tools for calculating costs of constructs

Reply:
Can we add a place holder for this? I think its something we need to keep in mind as we develop, so it will be good to have a reminder and also for soliciting feedback from others.

Reply:
There is one already...